lib/src/local_working_copy.rs

```diff
@@ -1513,14 +1514,43 @@ impl FileSnapshotter<'_> {
             let id = self.write_file_to_store(repo_path, disk_path).await?;
             // On Windows, we preserve the executable bit from the current tree.
             let executable = executable.unwrap_or_else(|| {
-                if let Some(TreeValue::File { id: _, executable }) = current_tree_value {
+                if let Some(TreeValue::File {
+                    id: _,
+                    executable,
+                    copy_id: _,
+                }) = current_tree_value
+                {
                     *executable
                 } else {
                     false
                 }
             });
-            Ok(Merge::normal(TreeValue::File { id, executable }))
+            // Preserve the copy id from the current tree
+            let copy_id = {
+                if let Some(TreeValue::File {
+                    id: _,
+                    executable: _,
+                    copy_id,
+                }) = current_tree_value
+                {
+                    copy_id.clone()
+                } else {
+                    CopyId::placeholder()
+                }
+            };
+            Ok(Merge::normal(TreeValue::File {
+                id,
+                executable,
+                copy_id,
+            }))
         } else if let Some(old_file_ids) = current_tree_values.to_file_merge() {
+            // Safe to unwrap because the copy id exists exactly on the file variant
+            let copy_id_merge = current_tree_values.to_copy_id_merge().unwrap();
+            let copy_id = copy_id_merge
+                .resolve_trivial()
+                .cloned()
+                .flatten()
+                .unwrap_or_else(CopyId::placeholder);
```

Comment:
Just a comment. We'll need to handle unresolved case and resolved to `None` differently. There may be change-delete content conflict `A - B + 0` with unchanged copy history `X - X + 0`. Perhaps, `resolve_file_executable()` can be generic over `Merge<Option<T>>`.

This can be addressed later.